src/main/java/com/example/Session.java

```diff
@@ -0,0 +1,137 @@
+package com.example;
+
+import com.example.annotation.Column;
+import com.example.annotation.Id;
+import com.example.annotation.Table;
+import lombok.SneakyThrows;
+
+import javax.sql.DataSource;
+import java.lang.reflect.Field;
+import java.sql.Connection;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+public class Session {
+    private final DataSource dataSource;
+    private final Map<EntityKey, Object> entityCacheMap;
+    private final Map<EntityKey, Map<String, Object>> entitySnapshotMap;
+
+    public Session(DataSource dataSource) {
+        this.dataSource = dataSource;
+        entityCacheMap = new HashMap<>();
+        entitySnapshotMap = new HashMap<>();
+    }
+
+    @SneakyThrows
+    public <T> T find(Class<T> entityType, Object id) {
+        EntityKey entityKey = new EntityKey(id, entityType);
+        if (entityCacheMap.containsKey(entityKey)) {
+            return entityType.cast(entityCacheMap.get(entityKey));
+        }
+        try (Connection connection = dataSource.getConnection()) {
+            String tableName = getTableName(entityType);
+            String idColumnName = getColumnName(getIdField(entityType));
+            String selectByIdQuery = "SELECT * FROM %s WHERE %s = ?" .formatted(tableName, idColumnName);
+            try (PreparedStatement preparedStatement = connection.prepareStatement(selectByIdQuery)) {
+                preparedStatement.setObject(1, id);
+                ResultSet resultSet = preparedStatement.executeQuery();
+                T entity = mapResultSetToEntity(resultSet, entityType);
+                entityCacheMap.put(entityKey, entity);
+                entitySnapshotMap.put(entityKey, createFieldNameToValueMap(entity));
+                return entity;
+            }
+        }
+    }
+
+    @SneakyThrows
+    private static <T> Map<String, Object> createFieldNameToValueMap(T entity) {
+        Map<String, Object> map = new HashMap<>();
+        for (Field field : entity.getClass().getDeclaredFields()) {
+            Object value = field.get(entity);
```

Comment:
Potential `IllegalAccessException` here, `field.setAccessible(true)` should be used for cases of private fields